src/test/java/main/java/com/google/googleidentity/security/LoginCheckServletTest.java

```diff
@@ -0,0 +1,134 @@
+    limitations under the License.
+*/
+
+package main.java.com.google.googleidentity.security;
+
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.logging.Logger;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.common.base.Charsets;
+import com.google.common.hash.Hashing;
+import com.google.googleidentity.filter.UserAuthenticationFilter;
+import com.google.googleidentity.security.LoginCheckServlet;
+import com.google.googleidentity.security.UserSession;
+import com.google.googleidentity.user.InMemoryUserDetailsService;
+import com.google.googleidentity.user.UserDetails;
+import com.google.googleidentity.user.UserDetailsService;
+import com.google.inject.Injector;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertFalse;
+
+import com.google.inject.Provider;
+
+import org.junit.Before;
+import org.junit.Test;
+
+
+import static org.mockito.Mockito.*;
+
+/**
+ * Test logincheckservelet's logic
+ *
+ */
+public class LoginCheckServletTest  {
+
+    public class UserSessionProvider implements Provider<UserSession> {
+        @Override
+        public UserSession get() {
+            UserSession userSession = new UserSession();
+            userSession.setUser(UserDetails.newBuilder()
+                    .setUsername("user")
+                    .setPassword(Hashing.sha256()
+                            .hashString("123456", Charsets.UTF_8).toString())
+                    .build());
+            return userSession;
+        }
+    }
+
+    @Test
+    public void testLoginCheckServlet_correctLoginRequest_redirectToResource()
+            throws ServletException, IOException {
+
+        UserDetailsService userDetailsService = new InMemoryUserDetailsService();
+
+        userDetailsService.addUser(UserDetails.newBuilder()
+                .setUsername("user")
+                .setPassword(Hashing.sha256()
+                        .hashString("123456", Charsets.UTF_8).toString())
+                .build());
+
+        LoginCheckServlet loginCheckServlet = new LoginCheckServlet(
+                new UserSessionProvider(), userDetailsService);
+
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
+
+        when(request.getParameter("username")).thenReturn("user");
+        when(request.getParameter("password")).thenReturn(Hashing.sha256()
+                .hashString("123456", Charsets.UTF_8).toString());
+
+        StringWriter stringWriter = new StringWriter();
+        PrintWriter writer = new PrintWriter(stringWriter);
+        when(response.getWriter()).thenReturn(writer);
+
+        loginCheckServlet.doPost(request, response);
+
+        assertTrue(stringWriter.toString().startsWith("/resource/user"));
```

Comment:
since there is no random things underlying, maybe we can check the full string value?